[exporter/azuremonitor] support sending to multiple azuremonitor exporter

hgaol opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

currently, azuremonitor exporter doesn't support sending to multiple application insights. This PR will add this feature.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #34188 

<!--Describe what testing was performed and which tests were added.-->
#### Testing


<!--Describe the documentation added.-->
#### Documentation

No additional configs.

<!--Please delete paragraphs that you did not use before submitting.-->

hgaol committed:
support sending to multiple azuremonitor exporter

hgaol committed:
add change log

hgaol committed:
fix CI

atoulme commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -101,43 +110,65 @@ func (f *factory) createMetricsExporter(
 }
 
+func (f *factory) initLogger(logger *zap.Logger) {
+	if f.hasInitLogger {
```
sync.Once seems like a better approach

hgaol replied:
udpated with sync.Once.

atoulme commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -63,7 +70,8 @@ func (f *factory) createTracesExporter(
 	}
 
-	tc, errInstrumentationKeyOrConnectionString := f.getTransportChannel(exporterConfig, set.Logger)
+	f.initLogger(set.Logger)
```
you cannot assume the same Logger is passed in for all. You might need to change this logic.

hgaol replied:
Yes, but the `appinsights` SDK only has one global diagnostics writer. If adding each logger to listener, there'll be dup messages. I don't know any elegant way to solve this issue at this point. any suggestion?

https://github.com/microsoft/ApplicationInsights-Go/blob/c063db145320a271b46e51eef7bceb88176b82c1/appinsights/diagnostics.go#L32C1-L33C52

```go
// The one and only diagnostics writer.
var diagnosticsWriter = &diagnosticsMessageWriter{}
```

atoulme commented:
It feels like the factory should not handle any of the channel business. The component can be reused with the sharedcomponent lib, which feels more adequate.

hgaol commented:
> It feels like the factory should not handle any of the channel business. The component can be reused with the sharedcomponent lib, which feels more adequate.

Currently in `azuremonitorexporter`, the metric, log, trace are separate exporters, and for same component they're using the same transport channel. IMO, shared component is not suitable for such case. I plan to merge metric ,log, trace exporter into single one exporter and use shared component lib to cache it. Just like what `fileexporter` do. It needs more code changes as well as tests.

hgaol committed:
using sync.Once for logger init

hgaol committed:
merge metric,log,trace expoter into azuremonitor exporter and use sharedcomponent to cache exporter by component id

hgaol committed:
bad, race

hgaol committed:
update test

hgaol committed:
fix race in test

hgaol commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -101,43 +110,65 @@ func (f *factory) createMetricsExporter(
+	if f.tChannels == nil {
+		f.tChannels = make(map[component.ID]transportChannel)
 	}
+	f.tChannels[id] = tChannel
```
updated. channel will be closed in `Shutdown` function of AzureMonitorExporter.

hgaol commented:
> > It feels like the factory should not handle any of the channel business. The component can be reused with the sharedcomponent lib, which feels more adequate.
> 
> Currently in `azuremonitorexporter`, the metric, log, trace are separate exporters, and for same component they're using the same transport channel. IMO, shared component is not suitable for such case. I plan to merge metric ,log, trace exporter into single one exporter and use shared component lib to cache it. Just like what `fileexporter` do. It needs more code changes as well as tests.

Hi @atoulme , I've done below changes and validated locally and it can ingest into multiple appinsights successfully. pls let me know if any comments, thanks for your careful review!
 - merged metric, log, trace exporters into `AzureMonitorExporter`, and deleted `metric_exporter`, `log_exporter` and `trace_exporter`.
 - used `sharedcomponent` to cache it by component id. 
 - add `start` and `shutdown` to initialize and close channel.
 - adjusted tests.

hgaol committed:
fix CI issue

hgaol commented:
Looks the failed checks are not caused by my change?

hgaol committed:
fix issues

hgaol commented:
bump, anyone can help reviewing this PR? thanks!

hgaol committed:
Merge branch 'main' into 34188

hgaol committed:
Merge branch 'main' into 34188

hgaol commented:
I've validated using 2 Azure appinsights, and the new change is compatible with old ones.

hgaol committed:
Merge branch 'main' into 34188

hgaol committed:
Merge branch 'main' into 34188

hgaol committed:
Merge branch 'main' into 34188

hgaol committed:
Merge branch 'main' into 34188

hgaol commented:
Gently ping, looks code owner hasn't noticed this PR yet, anyone from @open-telemetry/collector-contrib-approvers could help to take a look? Thanks! BTW, I'm glad to be the co-owner for this component for further triage/review to balance the load if possible.

TylerHelmuth commented:
@pcwiese, @hgaol please review as code owners. Failure to review may lead this component to being marked unmaintained.

hgaol commented:
Hi @TylerHelmuth , I saw you removed `waiting-for-code-owners` label. I was just added as owner of this component, and I'm also owner of this PR. Is there any process different? Should I still wait for the original owner (pcwiese) to review?

hgaol committed:
Merge branch 'main' into 34188

TylerHelmuth commented:
Ya I removed the label bc you're a code owner now and you submitted the pr. Now we need an approver to approve.

hgaol commented:
> Ya I removed the label bc you're a code owner now and you submitted the pr. Now we need an approver to approve.

Thanks for the explanation! I'll re-solve CI issues and let you know once done!

hgaol committed:
fix CI issue

hgaol commented:
Hi @TylerHelmuth , I've fixed CI issues and validated. could you or someone from approvers to take a look? thanks!

atoulme commented on exporter/azuremonitorexporter/config.go:
```diff
@@ -19,4 +19,5 @@ type Config struct {
 	MaxBatchSize       int                        `mapstructure:"maxbatchsize"`
 	MaxBatchInterval   time.Duration              `mapstructure:"maxbatchinterval"`
 	SpanEventsEnabled  bool                       `mapstructure:"spaneventsenabled"`
+	ShutdownTimeout    time.Duration              `mapstructure:"shutdown_timeout"`
```
please add to README

hgaol replied:
added

atoulme commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -8,23 +8,39 @@ package azuremonitorexporter // import "github.com/open-telemetry/opentelemetry-
+	exporters                      = sharedcomponent.NewSharedComponents()
+)
+
+type AzureMonitorExporter interface {
```
does this interface need to be exported?

hgaol replied:
removed

atoulme commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -52,87 +69,84 @@ func (f *factory) createTracesExporter(
+	ame := getOrCreateAzureMonitorExporter(cfg, set)
+
+	return exporterhelper.NewTraces(
+		context.TODO(),
```
you should use the context passed in line 68

hgaol replied:
resolved

atoulme commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -52,87 +69,84 @@ func (f *factory) createTracesExporter(
+	ame := getOrCreateAzureMonitorExporter(cfg, set)
+
+	return exporterhelper.NewLogs(
+		context.TODO(),
```
same for context

hgaol replied:
resolved

atoulme commented on exporter/azuremonitorexporter/factory.go:
```diff
@@ -52,87 +69,84 @@ func (f *factory) createTracesExporter(
+	ame := getOrCreateAzureMonitorExporter(cfg, set)
+
+	return exporterhelper.NewMetrics(
+		context.TODO(),
```
same for context

hgaol replied:
resolved

hgaol commented on exporter/azuremonitorexporter/azuremonitor_exporter.go:
```diff
@@ -0,0 +1,165 @@
+	logger           *zap.Logger
+	packer           *metricPacker
+	startOnce        sync.Once
+	endOnce          sync.Once
```
Yes, looks it'll still run `Start` 3 times for each shared component. Let me take some time to investigate.

hgaol replied:
I think it because that if I use one component in multiple pipelines, for example use it as exporter for `metrics/1`, `logs/1` and `traces/1`, there're 3 nodes and will trigger `component.Start` 3 times?
https://github.com/open-telemetry/opentelemetry-collector/blob/dae5d199ffc602b3d4633301584ad7d40ae14909/service/internal/graph/graph.go#L419

atoulme replied:
I think sharedcomponent takes care of this? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/sharedcomponent/sharedcomponent.go#L62

hgaol replied:
Thanks @atoulme ! You're right, updated using `sharedcomponent.Start` and remove `startOnce`, `endOnce`.

hgaol committed:
udpate for comments

hgaol committed:
fix

hgaol committed:
udpate

hgaol committed:
refactor: remove unnecessary sync.Once usage in azureMonitorExporter

hgaol committed:
Merge branch 'main' into 34188

atoulme commented on exporter/azuremonitorexporter/README.md:
```diff
@@ -42,6 +42,7 @@ The following settings can be optionally configured:
   - `num_consumers` (default = 10): Number of consumers that dequeue batches; ignored if `enabled` is `false`
   - `queue_size` (default = 1000): Maximum number of batches kept in memory before data; ignored if `enabled` is `false`
   - `storage` (default = `none`): When set, enables persistence and uses the component specified as a storage extension for the persistent queue
+- `shutdown_timeout` (default = 1s): When component is shutdowning, it'll force shutdown this component when exceed the timeout
```
```suggestion
- `shutdown_timeout` (default = 1s): Timeout to wait for graceful shutdown. Once exceeded, the component will shut down forcibly, dropping any element in queue.
```

hgaol committed:
Update exporter/azuremonitorexporter/README.md

Co-authored-by: Antoine Toulme <antoine@toulme.name>

atoulme committed:
Merge branch 'main' into 34188

atoulme reviewed (approved):
Changes LGTM. @hgaol note those changes are extensive and I hope I am not missing anything

hgaol commented:
Thanks @atoulme for your careful review! Agree that it's a big change. I've validated using both `connection_string` and `instrumentation_key`, with single or multiple application insights, looks it works well. I'll continue follow if any regression issue.
